src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1119,50 +1057,82 @@ void CASESession::SendErrorMsg(SigmaErrorType errorCode)
                    ChipLogError(SecureChannel, "Failed to send error message"));
 }
 
-CHIP_ERROR CASESession::FindValidTrustedRoot(const System::PacketBufferTLVReader & tlvReader, uint32_t nTrustedRoots)
+CHIP_ERROR CASESession::GenerateDestinationID(const ByteSpan & random, const Credentials::CertificateKeyId * trustedRootId,
+                                              NodeId nodeId, FabricId fabricId, MutableByteSpan & destinationId)
 {
-    CertificateKeyId trustedRoot;
-    System::PacketBufferTLVReader suppTlvReader;
-    uint8_t trustedRootId[kTrustedRootIdSize];
+    HMAC_sha hmac;
+    uint8_t destinationMessage[kDestinationMessageLen];
+    const ChipCertificateData * rootCertificate = mOpCredSet->GetRootCertificate(*trustedRootId);
+    VerifyOrReturnError(rootCertificate != nullptr, CHIP_ERROR_CERT_NOT_FOUND);
+    VerifyOrReturnError(!rootCertificate->mPublicKey.empty(), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(rootCertificate->mPublicKey.size() == kP256_PublicKey_Length, CHIP_ERROR_INTERNAL);
+
+    Encoding::LittleEndian::BufferWriter bbuf(destinationMessage, sizeof(destinationMessage));
 
-    trustedRoot = CertificateKeyId(trustedRootId);
+    bbuf.Put(random.data(), random.size());
+    bbuf.Put(rootCertificate->mPublicKey.data(), rootCertificate->mPublicKey.size());
+    bbuf.Put64(fabricId);
+    bbuf.Put64(nodeId);
+
+    VerifyOrReturnError(bbuf.Fit(), CHIP_ERROR_NO_MEMORY);
 
-    for (uint32_t i = 0; i < nTrustedRoots; ++i)
+    // retrieve Fabric IPK
```

Comment:
IPK should be passed as an argument. The loop in `FindDestinationIdCandidate()` forgets to iterate over all IPKs. At least if the list of IPKs, instead of coming from a crypto store API, came from a local list in  `FindDestinationIdCandidate()`, with a TODO to actually populate w/ each IPK for the given fabric, it would allow moving to "fully correct" impl with just changing how that IPK list is obtained, rather than chaning this method structurally.

Reply:
Updated.